### PR TITLE
[Keyboard] Add KMAC PAD

### PR DESCRIPTION
**Pull request by talsu**
## Description

Implemented to enable use of KMAC PAD released with KMAC keyboard.
QMK already support KMAC, but does not support KMAC PAD.
The default keymap implements the same functionality as the official default KMAC PAD firmware.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/kmac_pad/rules.mk [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+# Bootloader selection
+BOOTLOADER = atmel-dfu
+
+# Build Options
+BOOTMAGIC_ENABLE = yes      # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no        # Commands for debug and configuration
+CUSTOM_MATRIX = yes         # Custom matrix file
```
Comment (drashna) [inline keyboards/kmac_pad/rules.mk:16]: Could this be converted to use "lite"?

Comment (talsu) [reply keyboards/kmac_pad/rules.mk]: Yes, I have changed. I liked that the code was simplified. Thank you. 😄

File: keyboards/kmac_pad/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    switch (keycode) {
+        case MD_BOOT:
+            if (record->event.pressed) {
+                key_timer = timer_read32();
+            } else {
+                if (timer_elapsed32(key_timer) >= 2000) {
+                    reset_keyboard();
+                }
+            }
+            return false;
```
Comment (drashna) [inline keyboards/kmac_pad/keymaps/default/keymap.c:65]:
```suggestion

    switch (keycode) {
        case MD_BOOT:
        {
            static uint32_t key_timer;
            if (record->event.pressed) {
                key_timer = timer_read32();
            } else {
                if (timer_elapsed32(key_timer) >= 2000) {
                    reset_keyboard();
                }
            }
            return false;
        }
```

File: keyboards/kmac_pad/kmac_pad.c [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+void keyboard_pre_init_kb(void) {
+
+    /* Set Backlight pin as output
+     * FN Pin PB3
+     * PAD Pin PB1
+     */
+    setPinOutput(B3);
+    setPinOutput(B1);
+}
```
Comment (drashna) [inline keyboards/kmac_pad/kmac_pad.c:28]:
```suggestion
    keyboard_pre_init_user();
}
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Review by drashna — APPROVED** [review]

File: keyboards/kmac_pad/config.h [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+//#define MAGIC_KEY_LAYER6         6
+//#define MAGIC_KEY_LAYER7         7
+//#define MAGIC_KEY_LAYER8         8
+//#define MAGIC_KEY_LAYER9         9
+//#define MAGIC_KEY_BOOTLOADER     PAUSE
+//#define MAGIC_KEY_LOCK           CAPS
+//#define MAGIC_KEY_EEPROM         E
+//#define MAGIC_KEY_NKRO           N
+//#define MAGIC_KEY_SLEEP_LED      Z
+
```
Comment (fauxpark) [inline keyboards/kmac_pad/config.h:133]:
```suggestion
```

File: keyboards/kmac_pad/config.h [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+/* enable advanced MIDI features:
+   - MIDI notes can be added to the keymap
+   - Octave shift and transpose
+   - Virtual sustain, portamento, and modulation wheel
+   - etc.
+*/
+//#define MIDI_ADVANCED
+
+/* override number of MIDI tone keycodes (each octave adds 12 keycodes and allocates 12 bytes) */
+//#define MIDI_TONE_KEYCODE_OCTAVES 1
```
Comment (fauxpark) [inline keyboards/kmac_pad/config.h:170]:
```suggestion
```

File: keyboards/kmac_pad/info.json [hunk 6]
```diff
@@ -0,0 +1,36 @@
+{
+  "keyboard_name": "KMAC PAD",
+  "maintainer": "talsu",
+  "width": 4,
+  "height": 6.5,
+  "layouts": {
+    "LAYOUT": {
+      "key_count": 18,
```
Comment (fauxpark) [inline keyboards/kmac_pad/info.json:8]:
```suggestion
```

File: keyboards/kmac_pad/readme.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
+# KMAC PAD
+
+![kmac_pad](https://i.imgur.com/4P1ybgNl.jpg)
+
+KMAC PAD is a num pad keyboard.
+It can be used independently, but can also be used by connecting with KMAC keyboard case.
+
+Keyboard Maintainer: [talsu](https://github.com/talsu)  
+Hardware Supported: KMAC PAD  
+Hardware Availability: http://www.kbdmania.net/xe/news/5232321
```
Comment (fauxpark) [inline keyboards/kmac_pad/readme.md:10]:
```suggestion
* Keyboard Maintainer: [talsu](https://github.com/talsu)
* Hardware Supported: KMAC PAD
* Hardware Availability: http://www.kbdmania.net/xe/news/5232321
```

File: keyboards/kmac_pad/rules.mk [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
```
Comment (fauxpark) [inline keyboards/kmac_pad/rules.mk:27]:
```suggestion
# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = full     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
CUSTOM_MATRIX = lite        # Custom matrix file
```

File: keyboards/kmac_pad/matrix.c [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+#include <stdint.h>
+#include <stdbool.h>
+#include "wait.h"
+#include "print.h"
+#include "debug.h"
+#include "util.h"
+#include "matrix.h"
+#include "debounce.h"
+#include "quantum.h"
```
Comment (fauxpark) [inline keyboards/kmac_pad/matrix.c:26]:
```suggestion
#include "matrix.h"
#include "quantum.h"
```

File: keyboards/kmac_pad/matrix.c [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <stdint.h>
+#include <stdbool.h>
```
Comment (fauxpark) [inline keyboards/kmac_pad/matrix.c:19]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!

**Pull request merged**